Overhaul ECALL / Syscall handling

leekillough opened the pull request:
1. It uses `__attribute__((naked))` to prevent the compiler from generating any prologue/epilogue code in the Syscalls wrappers, or from making any assumptions about the uses of the Syscall wrapper's arguments, or of how the return value register is set. This allows raw inline assembly, and we set `a7` and then do the `ECALL` directly, and then `RET`, since the `a0` value returned from `ECALL` is the return register for the wrapper. <br> <br>**Previously, at `-O3` the entire wrapper call could be eliminated (not just inlined) by the compiler, because there was no indication that it used its arguments, or that it generated its return value from inline assembly.**<br><br>It would be very messy to do this the conventional way for each Syscall, telling the compiler how its function parameters are passed into the inline assembly as inputs, making sure they are placed in the correct registers for the `ECALL`, and telling the compiler how the `ECALL` return register of `a0` needs to become the wrapper function's return register `a0` so that it doesn't think it can just throw away a call to the wrapper. With `__attribute__((naked))`, this complexity is avoided, and the wrappers can become one-liners.
2. It uses `static` for the wrapper functions, since this is in a header file and may be `#include`d more than once. `inline` is incompatible with `__attribute__((naked))`, so `inline` was left out. <br><br>**Previously, multiply defined symbols could occur if the same Syscall wrapper function was called from two different `.c` files.**
3. It uses a macro to simplify the generation of the wrappers, and gets rid of a redundant "table of comments". The macro takes the Syscall number and the prototype of the wrapper as its two arguments.
4. It fixes the incorrect names of some Syscalls in `include/RevProc.h`'s comments.<br>
5. It adds parameter names to C function _definitions_, because **parameter names were not optional in function _definitions_ in C until C20, and they produce a lot of warnings.** In particular, the `socket` Syscalls did not have parameter names.<br>
6. It fixes the return types of many of the Syscalls, since simply using `int` will cause C/C++ code to expect a 32-bit result, and it will be sign-extended on RV64. `long`, `size_t`, `ssize_t`, `void*`, and other types should be used instead of always using `int`. **An automated program going through Syscall prototypes in headers/manpages should be eventually used to determine the correct return types.**
7. It waits for `a0`-`a7` dependencies to be satisfied before executing `ECALL`. Although different Syscalls take different arguments, waiting for `a0`-`a7` to be ready is general enough for any Syscall. Fixes #253.
8. ECALLs are implemented by setting SCAUSE during the `ECALL` instruction, and then at each `ClockTick`, checking whether SCAUSE is set, and if so, executing the Ecall handler and stalling the current instruction. SCAUSE is cleared when the Ecall handler indicates that it has completed. There is no backtracking of `PC` anymore, and the `ECALL` instruction only gets executed once.
9. `SCAUSE` is implemented as a simple `enum class` member of `RevRegFile`. There are no longer two versions of it, one for RV32 and one for RV64.
10. The `ECALL_*` Syscall handler functions no longer take a `RevInst` parameter, since it was never used.
11. The `Ecalls` table mapping Syscall numbers to handler functions, maps them to `RevProc` member function pointers, rather than the much more expensive `std::function`.
12. The `Ecalls` table is `static` now, and is initialized in `src/RevSysCalls.cc`.
13. `EXCEPTION_CAUSE` in `RevInstTable.h` is renamed to `RevExceptionCause` in `RevRegFile.h` and is made an `enum class`. This makes it more type-safe and abides by the CamelCase naming convention.
14. `EcallStatus`'s values `SUCCESS`, `CONTINUE`, `ERROR` are no longer defined in terms of `EXCEPTION_CAUSE`, because the exception cause is unrelated to the status of an `ECALL` handler function.

rkabrick reviewed (approved):
Looks much better than it used to! Nice work.

leekillough committed:
Fix syscall wrappers

    Use __attribute__((naked)) to avoid bad optimization of syscall wrappers

    Use static to avoid multiple definitions of syscall wrappers

leekillough committed:
test hunch

leekillough committed:
another test

rkabrick committed:
Clear Ecall state every time we successfully execute an ecall

leekillough committed:
Merge remote-tracking branch 'tcl/devel' into fix_syscall_wrappers

leekillough committed:
update

leekillough committed:
update

leekillough committed:
Merge remote-tracking branch 'tcl/devel' into fix_syscall_wrappers

leekillough committed:
remove unknown warning

leekillough committed:
Merge remote-tracking branch 'tcl/devel' into fix_syscall_wrappers

leekillough committed:
Add ecall dependency checking

leekillough committed:
Make ECALLs work without rewinding PC, add dependency checks for ECALL

leekillough committed:
handle EcallStatus::ERROR correctly

leekillough committed:
change logic

leekillough committed:
fix failures

leekillough committed:
revert change to rev_write test

leekillough committed:
rename SYSCALL to REV_SYSCALL to reduce the chances of conflict

jleidel reviewed (approved):
good to merge

leekillough committed:
Merge remote-tracking branch 'tcl/devel' into fix_syscall_wrappers

donofrio reviewed (approved):
Looks good - this should address a bunch of bugs
